packages/loader/container-loader/src/test/serializedStateManager.spec.ts

```diff
@@ -206,55 +237,309 @@ describe("serializedStateManager", () => {
+			{ notifyImminentClosure: false },
+			"clientId",
+			new MockRuntime(),
+			resolvedUrl,
+		);
+		const parsed = JSON.parse(state) as IPendingContainerState;
+		assert.strictEqual(parsed.baseSnapshot.id, "fromStorage");
+		const attributes = getAttributesFromPendingState(parsed);
+		assert.strictEqual(
+			attributes.sequenceNumber,
+			snapshotSequenceNumber,
+			"wrong snapshot sequence number",
+		);
+		assert.strictEqual(
+			parsed.savedOps[0].sequenceNumber,
+			snapshotSequenceNumber + 1,
+			"wrong first saved op",
+		);
+		assert.strictEqual(
+			parsed.savedOps[parsed.savedOps.length - 1].sequenceNumber,
+			processedOpsSize,
+			"wrong last saved op",
+		);
+	});
+
+	it("refresh snapshot when there are no processed ops", async () => {
+		const pending: IPendingContainerState = {
+			...pendingLocalState,
+			baseSnapshot: { ...snapshot, id: "fromPending" },
+		};
+		const storageAdapter = new MockStorageAdapter();
+		const deferred = new Deferred<void>();
+		const cb = () => {
+			deferred.resolve();
+		};
+		const serializedStateManager = new SerializedStateManager(
+			pending,
+			logger.toTelemetryLogger(),
+			storageAdapter,
+			true,
+			cb,
+		);
+		const snapshotSequenceNumber = 11;
+		storageAdapter.uploadSummary(snapshotSequenceNumber);
+
+		await serializedStateManager.fetchSnapshot(undefined, false);
+		await deferred.promise;
+		const state = await serializedStateManager.getPendingLocalStateCore(
+			{ notifyImminentClosure: false },
+			"clientId",
+			new MockRuntime(),
+			resolvedUrl,
+		);
+		const parsed = JSON.parse(state) as IPendingContainerState;
+		assert.strictEqual(parsed.baseSnapshot.id, "fromStorage");
+		const attributes = getAttributesFromPendingState(parsed);
+		assert.strictEqual(
+			attributes.sequenceNumber,
+			snapshotSequenceNumber,
+			"wrong snapshot sequence number",
+		);
+		assert.strictEqual(parsed.savedOps.length, 0, "should be no savedOps");
+	});
+
+	it("refresh snapshot when snapshot sequence number is above processed ops", async () => {
+		const pending: IPendingContainerState = {
+			...pendingLocalState,
+			baseSnapshot: { ...snapshot, id: "fromPending" },
+		};
+		const storageAdapter = new MockStorageAdapter();
+		const deferred = new Deferred<void>();
+		const cb = () => {
+			deferred.resolve();
+		};
+		const serializedStateManager = new SerializedStateManager(
+			pending,
+			logger.toTelemetryLogger(),
+			storageAdapter,
+			true,
+			cb,
+		);
+		const processedOpsSize1 = 20;
+		for (let num = 0; num < processedOpsSize1; ++num) {
+			serializedStateManager.addProcessedOp(generateSavedOp());
+		}
+
+		const snapshotSequenceNumber = 22;
+		storageAdapter.uploadSummary(snapshotSequenceNumber);
+
+		await serializedStateManager.fetchSnapshot(undefined, false);
+		const processedOpsSize2 = 10;
+		for (let num = 0; num < processedOpsSize2; ++num) {
+			serializedStateManager.addProcessedOp(generateSavedOp());
+		}
```

Comment:
this is really hard to follow. i think changing generateSavedOp to take a sequence number would make it easier, and then the test would manage the sequence numbers explicitly, rather than implicitly via counts
```suggestion
		let seq = 0;
		while(seq <20){
		     serializedStateManager.addProcessedOp(generateSavedOp(seq++));
		}
		storageAdapter.uploadSummary(seq);

		await serializedStateManager.fetchSnapshot(undefined, false);
		
		while(seq <10){
		     serializedStateManager.addProcessedOp(generateSavedOp(seq++));
		}
```

Reply:
this has the added advantage of getting ride of shared state, which makes code make more complicated to maintain.